### PR TITLE
Use Typed Buffers in Arrays (#1811) (#1176)

### LINKED ISSUE #1811: Replace RawPtrBox with Safe Abstraction
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently the Array implementations make use of `RawPtrBox` to reference data on their attached ArrayData. Not only is this construction deeply unsafe, but requires further unsafe pointer manipulation to use.

**Describe the solution you'd like**

I would like for arrays to store strongly typed slices, instead of strongly typed pointers. This necessarily implies a self-referential struct, fortunately [ouroboros](https://crates.io/crates/ouroboros) provides a safe interface that handles this complexity for you. The hardest part is actually spelling the crates' name...

```
#[self_referencing]
pub struct PrimitiveArray<T: ArrowPrimitiveType> {
    data: ArrayData,
    #[borrows(data)]
    values: &'this [T::Native]
}
```

This would eliminate a lot of unsafe, without requiring any breaking changes.

**Describe alternatives you've considered**

We could not do this, this might just be a way to reduce the amount of unsafe in arrow-rs without requiring any breaking changes.

We could also abstract the self-referential shenanigans in a typed buffer abstraction, something like

```
#[self_referencing]
struct TypedBuffer<T: ArrowNativeType> {
  inner: Buffer,
  #[borrows(inner)]
  values: &'this [T]
}
```

Possibly even dropping ouroborous in favor of a little bit of custom unsafe. 

**Related Context**

Somewhat related to https://github.com/apache/arrow-rs/issues/1799 which introduces stronger typing within ArrayData itself.

We used ouroboros for a while in IOx to handle flatbuffers, and it did the job.  We have since removed flatbuffers in favor of a columnar protobuf representation, but that was more a reflection on flatbuffers than ouroboros.

Thoughts @alamb @jhorstmann @jorgecarleitao @HaoYang670 ?

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #1176
Closes #1811

# Rationale for this change
 

The first part of moving towards a fully typed ArrayData as part of #1176

As an added bonus the change to precompute the offset for `ByteArray` improves the performance of the string comparison kernels by up to 25%. The arithmetic kernels show a very slight performance regression of ~1-2%, I'm inclined to think this does not matter, we are talking microseconds here.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-buffer/src/buffer/immutable.rs [hunk 1]
```diff
@@ -162,12 +183,15 @@ impl Buffer {
     /// Panics iff `(offset + length)` is larger than the existing length.
     pub fn slice_with_length(&self, offset: usize, length: usize) -> Self {
         assert!(
-            offset + length <= self.len(),
+            offset.saturating_add(length) <= self.length,
```
Comment (tustvold) [inline arrow-buffer/src/buffer/immutable.rs:186]: This was actually a bug, now covered by https://github.com/apache/arrow-rs/pull/3743/files#diff-24dc7184f64fe7a484db137ef61b2ec31090bb959fc9013beefea7865cdfea59R628

File: arrow-array/src/record_batch.rs [hunk 2]
```diff
@@ -603,7 +603,7 @@ mod tests {
         let record_batch =
             RecordBatch::try_new(Arc::new(schema), vec![Arc::new(a), Arc::new(b)])
                 .unwrap();
-        assert_eq!(record_batch.get_array_memory_size(), 592);
+        assert_eq!(record_batch.get_array_memory_size(), 640);
```
Comment (tustvold) [inline arrow-array/src/record_batch.rs:606]: There is a slight regression in the size of the arrays themselves, typically they are passed around as references so I am inclined to think this doesn't really matter, it is also likely temporary

File: arrow-buffer/src/buffer/scalar.rs [hunk 3]
```diff
@@ -48,39 +46,50 @@ impl<T: ArrowNativeType> ScalarBuffer<T> {
...
 
+    #[inline]
     fn deref(&self) -> &Self::Target {
-        // SAFETY: Bounds checked in constructor and ptr is valid for the lifetime of self
-        unsafe { std::slice::from_raw_parts(self.ptr, self.len) }
+        // SAFETY: Verified alignment in From<Buffer>
+        unsafe {
+            std::slice::from_raw_parts(
+                self.buffer.as_ptr() as *const T,
+                self.buffer.len() / std::mem::size_of::<T>(),
```
Comment (tustvold) [inline arrow-buffer/src/buffer/scalar.rs:64]: I found caching this value doesn't impact performance, this is likely because the "hot" codepaths are either iterating the slice amortising this cost over multiple values, or using unchecked addressing where this length isn't ever consulted

File: arrow-array/src/array/mod.rs [hunk 4]
```diff
@@ -637,7 +637,7 @@ pub fn new_null_array(data_type: &DataType, length: usize) -> ArrayRef {
 }
 
 // Helper function for printing potentially long arrays.
-pub(crate) fn print_long_array<A, F>(
+fn print_long_array<A, F>(
```
Comment (tustvold) [inline arrow-array/src/array/mod.rs:640]: This is just a drive-by cleanup

File: arrow-buffer/src/buffer/immutable.rs [hunk 5]
```diff
@@ -30,26 +30,41 @@ use super::MutableBuffer;
...
     /// the internal byte buffer.
     data: Arc<Bytes>,
 
-    /// The offset into the buffer.
-    offset: usize,
+    /// Pointer into `data` valid
+    ///
+    /// We store a pointer instead of an offset to avoid pointer arithmetic
+    /// which causes LLVM to fail to vectorise code correctly
+    ptr: *const u8,
```
Comment (tustvold) [inline arrow-buffer/src/buffer/immutable.rs:42]: This seemingly inconsequential change makes a substantial difference to the benchmarks, LLVM seems to fail to realise that the offset doesn't change in the body of the loop and so fails to lift it out, which in turn causes the loop itself to fail to vectorise correctly

**Review by viirya — COMMENTED** [review]
Thank you @tustvold for moving this forward! I plan to review this in next few days.

File: arrow-array/src/array/list_array.rs [hunk 6]
```diff
@@ -243,14 +224,20 @@ impl<OffsetSize: OffsetSizeTrait> GenericListArray<OffsetSize> {
...
+            false => {
+                let buffer = ScalarBuffer::new(
+                    data.buffers()[0].clone(),
+                    data.offset(),
+                    data.len() + 1,
+                );
+                // Safety:
+                // ArrayData is valid
+                unsafe { OffsetBuffer::new_unchecked(buffer) }
+            }
```
Comment (viirya) [inline arrow-array/src/array/list_array.rs:238]: This seems occurring more than once. Maybe we can have a function for it?

File: arrow-buffer/src/buffer/offset.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::buffer::ScalarBuffer;
+use crate::{ArrowNativeType, MutableBuffer};
+use std::ops::Deref;
+
+/// A non-empty buffer of monotonically increasing, positive integers
+#[derive(Debug, Clone)]
+pub struct OffsetBuffer<O: ArrowNativeType>(ScalarBuffer<O>);
```
Comment (viirya) [inline arrow-buffer/src/buffer/offset.rs:24]:
```suggestion
pub struct OffsetBuffer<O: OffsetSizeTrait>(ScalarBuffer<O>);
```

File: arrow-buffer/src/buffer/scalar.rs [hunk 8]
```diff
@@ -48,39 +46,50 @@ impl<T: ArrowNativeType> ScalarBuffer<T> {
...
-        assert!(
-            prefix.is_empty() && suffix.is_empty(),
-            "buffer is not aligned to {size} byte boundary"
-        );
-
-        let ptr = offsets.as_ptr();
-        Self { buffer, ptr, len }
+        let byte_offset = offset.checked_mul(size).expect("offset overflow");
+        let byte_len = len.checked_mul(size).expect("length overflow");
+        buffer.slice_with_length(byte_offset, byte_len).into()
```
Comment (viirya) [inline arrow-buffer/src/buffer/scalar.rs:51]: No need to check alignment anymore?

**Review by viirya — APPROVED** [review]

File: arrow-buffer/src/buffer/scalar.rs (hunk 8 above)
Comment (tustvold) [reply arrow-buffer/src/buffer/scalar.rs:51]: The alignment gets checked in the From conversion

File: arrow-buffer/src/buffer/offset.rs (hunk 7 above)
Comment (tustvold) [reply arrow-buffer/src/buffer/offset.rs:24]: `OffsetSizeTrait` is defined in arrow-array and so can't be used here

File: arrow-array/src/array/mod.rs [hunk 9]
```diff
@@ -636,7 +638,28 @@ pub fn new_null_array(data_type: &DataType, length: usize) -> ArrayRef {
     make_array(ArrayData::new_null(data_type, length))
 }
 
-// Helper function for printing potentially long arrays.
+/// Helper function that gets offset from an [`ArrayData`]
+///
+/// # Safety
+///
+/// - ArrayData must contain a valid [`OffsetBuffer`] as its first buffer
+unsafe fn get_offsets<O: ArrowNativeType>(data: &ArrayData) -> OffsetBuffer<O> {
```
Comment (tustvold) [inline arrow-array/src/array/mod.rs:646]: This will eventually get removed as we make ArrayData an enumeration of typed buffers
